File: frontend/src/components/settings/settingsAttention.ts
/**
 * 设置红点判定（与 backend/app/settings_attention.py 对齐）。
 * 改规则时请同步改两端测试。
 */

/** 与 candidate.PLACEHOLDER_API_KEYS / is_llm_api_key_configured 对齐 */
const PLACEHOLDER_API_KEYS = new Set(["", "sk-none", "sk-your-key"]);

export function isDraftApiKeyConfigured(
  apiKey?: string,
  apiKeyMasked?: string,
): boolean {
  const key = (apiKey || "").trim();
  if (key && !PLACEHOLDER_API_KEYS.has(key) && !key.includes("***")) {
    return true;
  }
  // 已保存脱敏：占位密钥不会以 *** 掩码回显
  const masked = (apiKeyMasked || "").trim();
  if (!masked || PLACEHOLDER_API_KEYS.has(masked)) return false;
  return masked.includes("***") || masked === "****";
}

/** 草稿态：链上至少一条具备 model + base_url + 有效 Key。 */
export function draftChainNeedsSetup(
  candidates: Array<{
    model?: string;
    base_url?: string;
    api_key?: string;
    api_key_masked?: string;
  }>,
): boolean {
  return !candidates.some((c) => {
    const model = (c.model || "").trim();
    const base = (c.base_url || "").trim();
    if (!model || !base) return false;
    return isDraftApiKeyConfigured(c.api_key, c.api_key_masked);
  });
}

/** 价目行是否缺必填单价（与后端 price_row_needs_setup 对齐）。Cache 可选。 */
export function priceRowNeedsSetup(row: {
  model: string;
  prompt_per_1m: number | null;
  completion_per_1m: number | null;
  embed_per_1m: number | null;
  kinds?: string[];
}): boolean {
  const kinds = row.kinds ?? [];
  let hasEmbed = kinds.includes("embed");
  let hasChat = kinds.some((k) => k !== "embed");
  if (!hasEmbed && !hasChat) {
    const embedLike = /embed/i.test(row.model);
    hasEmbed = embedLike;
    hasChat = !embedLike;
  }
  if (hasChat && (row.prompt_per_1m == null || row.completion_per_1m == null)) {
    return true;
  }
  if (hasEmbed && row.embed_per_1m == null) {
    return true;
  }
  return false;
}

/** 面板打开时用草稿/本地态覆盖服务端对应分区，避免红点不同步。 */
export function mergeSettingsAttention(
  server: {
    any: boolean;
    model: { any: boolean; chat: boolean; utility: boolean; embed: boolean };
    memory: { any: boolean; pending_count: number };
    usage: { any: boolean; incomplete_price_count: number };
  },
  overlay?: {
    model?: { chat: boolean; utility: boolean; embed: boolean } | null;
    memoryPending?: number | null;
    usageIncomplete?: number | null;
  } | null,
): typeof server {
  const model = overlay?.model
    ? {
        chat: overlay.model.chat,
        utility: overlay.model.utility,
        embed: overlay.model.embed,
        any:
          overlay.model.chat || overlay.model.utility || overlay.model.embed,
      }
    : server.model;
  const pending =
    overlay?.memoryPending != null
      ? overlay.memoryPending
      : server.memory.pending_count;
  const incomplete =
    overlay?.usageIncomplete != null
      ? overlay.usageIncomplete
      : server.usage.incomplete_price_count;
  const memory = {
    any: pending > 0,
    pending_count: pending,
  };
  const usage = {
    any: incomplete > 0,
    incomplete_price_count: incomplete,
  };
  return {
    any: model.any || memory.any || usage.any,
    model,
    memory,
    usage,
  };
}

File: backend/app/settings_attention.py
"""主界面「需要处理」红点：模型必填链、记忆待确认、未填价目。"""

from __future__ import annotations

from typing import Any

from app.models.candidate import ModelChain, resolve_chain_candidates
from app.settings_store import is_llm_api_key_configured


def chain_needs_setup(settings: Any, chain: ModelChain) -> bool:
    """链上至少一条具备 model + base_url + 有效 api_key 才算已配置。"""
    for c in resolve_chain_candidates(settings, chain):
        if not (c.model or "").strip():
            continue
        if not (c.base_url or "").strip():
            continue
        if not is_llm_api_key_configured(c.api_key):
            continue
        return False
    return True


def price_row_needs_setup(row: dict[str, Any]) -> bool:
    """见过的模型若缺必填单价字段则需配置。Cache 可选。"""
    kinds = row.get("kinds") or []
    if isinstance(kinds, str):
        kinds = [k for k in kinds.split(",") if k]
    has_embed = "embed" in kinds
    has_chat = any(k != "embed" for k in kinds)
    if not has_embed and not has_chat:
        mid = str(row.get("model") or "").lower()
        has_embed = "embed" in mid
        has_chat = not has_embed
    if has_chat and (
        row.get("prompt_per_1m") is None or row.get("completion_per_1m") is None
    ):
        return True
    if has_embed and row.get("embed_per_1m") is None:
        return True
    return False


def count_incomplete_prices(prices: list[dict[str, Any]]) -> int:
    return sum(1 for row in prices if price_row_needs_setup(row))


def build_settings_attention(
    *,
    settings: Any,
    memory_pending_count: int = 0,
    incomplete_price_count: int = 0,
) -> dict[str, Any]:
    chat = chain_needs_setup(settings, "chat")
    utility = chain_needs_setup(settings, "utility")
    embed = chain_needs_setup(settings, "embed")
    model_any = chat or utility or embed
    memory_any = memory_pending_count > 0
    usage_any = incomplete_price_count > 0
    return {
        "any": model_any or memory_any or usage_any,
        "model": {
            "any": model_any,
            "chat": chat,
            "utility": utility,
            "embed": embed,
        },
        "memory": {
            "any": memory_any,
            "pending_count": int(memory_pending_count),
        },
        "usage": {
            "any": usage_any,
            "incomplete_price_count": int(incomplete_price_count),
        },
    }
